pkg/kubelet/server/streaming/request_cache.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package streaming
+
+import (
+	"container/list"
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util/clock"
+)
+
+var (
+	CacheTTL     = 5 * time.Minute
+	CacheMaxSize = 10000
+	TokenLen     = 24
+)
+
+type requestCache struct {
+	// clock is used to obtain the current time
+	clock clock.Clock
+
+	tokens map[string]*list.Element
+	ll     *list.List
+
+	lock sync.Mutex
+}
+
+// Type representing an *ExecRequest, *AttachRequest, or *PortForwardRequest.
+type request interface{}
+
+type cacheEntry struct {
+	token      string
+	req        request
+	expireTime time.Time
+}
+
+func newRequestCache() *requestCache {
+	return &requestCache{
+		clock:  clock.RealClock{},
+		ll:     list.New(),
+		tokens: make(map[string]*list.Element),
+	}
+}
+
+func (c *requestCache) startGC() {
+	gcTicker := c.clock.Tick(CacheTTL)
+	go func() {
+		for range gcTicker {
+			c.gc()
+		}
+	}()
+}
+
+// Insert the given request into the cache and returns the token used for fetching it out.
+func (c *requestCache) Insert(req request) (token string, err error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	token, err = c.uniqueToken()
+	if err != nil {
+		return "", err
+	}
+	ele := c.ll.PushFront(&cacheEntry{token, req, c.clock.Now().Add(CacheTTL)})
+	if c.ll.Len() > CacheMaxSize {
```

Comment:
> why not make this remove anything expired from the tail, instead of making it size-based?

fwiw, this will make insert operation O(n) in worst case, but on the other hand we can get rid of the extra gc goroutine.